proposals/20201026-creating-control-plane.md

```diff
@@ -0,0 +1,433 @@
+    NestedKAS NestedComponentSpec `json:"nestedKAS"`
+
+    // NestedKCM contains information required by the controller-manager provider to create the controller-manager
+    NestedKCM NestedComponentSpec `json:"nestedKCM"`
+}
+```
+
+We will let users specify component providers through NCP, implement a native mode that has each sub-controller to create the component on the super cluster, and define three new types `EtcdProvider`, `KASProvider`, and the `KCMProvider`:
+
+```go
+type KASProvider string
+
+type EtcdProvider string
+
+type KCMProvider string
+
+const (
+    NativeKASProvider KASProvider = "NativeKASProvider"
+
+    NativeEtcdProvider  EtcdProvider = "NativeEtcdProvider"
+    EtcdClusterOperator EtcdProvider = "EtcdClusterOperator"
+    EtcdOperator        EtcdProvider = "EtcdOperator"
+
+    NativeKCMProvider KCMProvider = "NativeKCMProvider"
+)
+```
+
+We will add three new fields `NestedEtcdProvider`, `NestedKASProvider`, and the `NestedKCMProvider` into the NCP, which will make the NCP look like the following:
+
+```go
+// NestedControlPlaneSpec defines the desired state of NestedControlPlane
+type NestedControlPlaneSpec struct {
+    // other fields.... 
+    
+    // NestedEtcdProvider specifies which EtcdProvider will be used to create the Etcd
+    // +optional
+    NestedEtcdProvider EtcdProvider `json:"etcdProvider,omitempty"`
+     
+    // NestedKASProvider specifies which KASProvider will be used to create the kube-apiserver
+    // +optional
+    NestedKASProvider KASProvider `json:"kasProvider,omitempty"`
+ 
+    // NestedKCMProvider specifies which KCMProvider will be used to create the kube-controller-manager
+    // +optional
+    NestedKCMProvider KCMProvider `json:"kcmProvider,omitempty"`
+}
+```
+
+### Bootstrap
+
+#### Create prerequisites 
+
+The NCP controller will need to create some resources before sub-controllers start creating the real NCP components.
+
+* **ClusterNamespace** is used to store all resources for the NCP.
+
+* **NKAS service**  is used to visit the NKAS. Since the IP/domain name is required when generating TLS certificates and kubeconfigs of NKAS. This service will need to be created in advance.
+
+* **NestedCRs**, including NEtcd, NKAS, and NKCM. are CRs that hold information required by the sub-controller. 
+
+After creating the above prerequisites, the sub-controller will start creating the workload of each component.
+
+As there exist dependencies between components, i.e., KAS cannot run without Etcd, KCM cannot work without KAS, when creating the NCP component, sub-controllers will need to get information and status of other CRs. Thereby, we will have each nested component CR points to the NCP CR. Moreover, To get other CRs through the NCP CR, we suggest adding object references of the three Nested CRs into the `NCP.status`, which will make the `NCP.status` look something like the following:
```

Comment:
If we keep CRs and NCP's in the same namespace we could use actual `metadata.ownerReferences`

Reply:
Are we going to assume that CRs and NCP will be located in the same namespace? As we intend to not create the `clusternamespace` for the user, should we introduce a new mechanism to ensures that the CRs and the NCP will be located in a same namespace?

Reply:
Yeah, `OwnerReferences` themselves will be forced to only work within the same namespace. (you can't associate to another NS). I think at this foundational level we should just expect that all CRs for a control plane `netcd, nas, ncm, ncp and nc (nested cluster)` are in the same namespace as where the components are provisioned.

Do you think that would cause any issues?

Reply:
No. I just worried that users may accidently put CRs in different ns, for example, forgetting to specify the ns and result in the CR being located in the default ns.